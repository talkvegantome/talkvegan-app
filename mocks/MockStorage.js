import { DateTime } from 'luxon';
export class MockStorage {
  constructor() {
    this.triggerUpdateMethods = [];
    this.settings = {
      language: 'en',
    };
    this.pageData = {
      en: {
        data: [
          {
            friendlyName: 'Test Page',
            rawContent: '# Test Page',
            relativePermalink: '/en/test-parent/test-page/',
            weight: null,
            permalink: 'https://talkveganto.me/en/test-parent/test-page/',
            section: {
              friendlyName: 'Test Parent',
              relativePermalink: '/en/test-parent/',
              weight: 1000,
            },
          },
        ],
        date: '2019-05-12T16:07:45.281077+01:00',
        lastSyncDate: DateTime.local()
          .plus({ days: -7 })
          .toISO(),
        languageName: 'English',
        languageShortCode: 'en',
      },
    };
  }
  addOnRefreshListener() {}
}
