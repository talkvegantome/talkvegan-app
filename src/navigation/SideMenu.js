import PropTypes from 'prop-types';
import React, {Component} from 'react';
import styles from '../styles/SideMenu.style.js';
import {ScrollView, View, SafeAreaView } from 'react-native';
import { ListItem, Divider } from 'react-native-elements';
import _ from 'lodash';

import Analytics from '../analytics'

import Pages from '../Pages.js'

class SideMenu extends Component {
  render () {
    return (
      <SafeAreaView style={styles.safeContainer}>
        <ScrollView style={styles.container} bounces={false}>
          <ListItem
            containerStyle={styles.navHeaderStyle}
            titleStyle={styles.navHeaderTitleStyle}
            // onPress={this.navigateToScreen('/'+this.state.settings.language+'/splash/')}
            title={this.props.title}/>
          <Divider style={{...styles.dividerStyle,...{height: 2}}} />
          <MenuItems {...this.props}/>
        </ScrollView>
      </SafeAreaView>
    );
  }
}

export class MenuItems extends Component{
  constructor(props) {
    super(props);
    this.props.storage.triggerUpdateMethods.push((storage) => {
      this.refreshStorage(storage)
    })
    let pages = new Pages(this.props.storage)
    let analytics = new Analytics(this.props.storage.settings)
    this.state = {
      analytics: analytics,
      settings: this.props.storage.settings,
      menu: pages.getMenu(),
      splashPath: pages.getSplashPath(),
      headerVisibility: {}
    };
  }

  refreshStorage(storage){
    let pages = new Pages(storage)
    let analytics = new Analytics(storage.settings)
    this.setState({
      analytics: analytics,
      settings: storage.settings,
      menu: pages.getMenu(storage),
      splashPath: pages.getSplashPath()
    })
  }
  navigateToScreen = (indexId) => () => {
    // Navigation is always to the 'Home' screen, but content changes based on the indexId
    this.props.navigation.navigate('Home', {indexId: indexId});
    this.props.navigation.closeDrawer();
    this.state.analytics.logEvent('navigateToPage', {relPath: indexId})
  }
  toggleHeaderVisibility = (headerName) => {
    let headerVisibility = this.state.headerVisibility
    if(headerVisibility[headerName]){
      headerVisibility[headerName] = false
    }else{
        _.forEach(headerVisibility, (val, headerName) => {
          headerVisibility[headerName] = false
        })
        headerVisibility[headerName] = true
    }

    this.setState({headerVisibility: headerVisibility})
  }

  generateMenuObjects(){
    let menuSorted = _.sortBy(this.state.menu, ['weight', 'friendlyName'])
    return _.map(menuSorted, (headerItem) => {
      let headerFriendlyName = headerItem.friendlyName
      let items = _.sortBy(headerItem.subItems, ['weight', 'friendlyName']).map((item) => {
        if(!_.isNil(item.displayInApp) && !item.displayInApp){
          // Don't display this page if it has displayInApp=false
          return 
        }
        return(
          <ListItem
            bottomDivider={true}
            containerStyle={styles.itemHeadingStyle}
            titleStyle={styles.itemHeadingTitleStyle}
            key={item.relativePermalink}
            onPress={this.navigateToScreen(item.relativePermalink)}
            title={item.friendlyName}
          />
        )
      })


      let headerVisibility = this.state.headerVisibility
      let display = headerFriendlyName in headerVisibility && headerVisibility[headerFriendlyName]? 'flex': 'none'
      return (
        <View key={headerFriendlyName}>
          <ListItem
            key={headerFriendlyName}
            containerStyle={styles.sectionHeadingStyle}
            titleStyle={styles.sectionHeadingTitleStyle}
            leftIcon={{
                name: display == 'flex' ? 'expand-less': 'expand-more', 
                iconStyle: styles.sectionHeadingTitleStyle
            }}
            title={headerFriendlyName}
            onPress={()=>{this.toggleHeaderVisibility(headerFriendlyName)}}
          />
          <Divider style={styles.dividerStyle} />
          <View style={{display:display}} animationType="slide">
            {items}
          </View>
        </View>
      )
    })
  }
  render(){
    return (
      <View>{this.generateMenuObjects()}</View>
    )
  }
}

SideMenu.propTypes = {
  navigation: PropTypes.object,
};

export default SideMenu;