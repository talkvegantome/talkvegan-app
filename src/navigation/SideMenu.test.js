import React from 'react';
import SideMenu from './SideMenu.js';
import renderer from 'react-test-renderer';

import {MockStorage} from '../../mocks/MockStorage.js'

let mockStorage = new MockStorage()

test('SideMenu Renders Correctly', () => {
  const tree = renderer.create(<SideMenu storage={mockStorage}/>).toJSON();
  expect(tree).toMatchSnapshot();
});
