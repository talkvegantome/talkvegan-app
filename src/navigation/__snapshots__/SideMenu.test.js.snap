// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideMenu Renders Correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#235789",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    style={
      Object {
        "backgroundColor": "#E8E9EB",
        "flex": 1,
      }
    }
  >
    <View>
      <View
        accessible={true}
        isTVSelectable={true}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#235789",
              "borderColor": "#bcbbc1",
              "flexDirection": "row",
              "padding": 14,
            }
          }
        >
          <View
            style={
              Object {
                "flex": 1,
                "justifyContent": "center",
              }
            }
          >
            <Text
              style={
                Object {
                  "backgroundColor": "transparent",
                  "color": "#E8E9EB",
                  "fontSize": 30,
                }
              }
              testID="listItemTitle"
              theme={
                Object {
                  "colors": Object {
                    "disabled": "hsl(208, 8%, 90%)",
                    "divider": "#bcbbc1",
                    "error": "#ff190c",
                    "grey0": "#393e42",
                    "grey1": "#43484d",
                    "grey2": "#5e6977",
                    "grey3": "#86939e",
                    "grey4": "#bdc6cf",
                    "grey5": "#e1e8ee",
                    "greyOutline": "#bbb",
                    "platform": Object {
                      "android": Object {
                        "error": "#f44336",
                        "primary": "#2196f3",
                        "secondary": "#9C27B0",
                        "success": "#4caf50",
                        "warning": "#ffeb3b",
                      },
                      "ios": Object {
                        "error": "#ff3b30",
                        "primary": "#007aff",
                        "secondary": "#5856d6",
                        "success": "#4cd964",
                        "warning": "#ffcc00",
                      },
                    },
                    "primary": "#2089dc",
                    "searchBg": "#303337",
                    "secondary": "#8F0CE8",
                    "success": "#52c41a",
                    "warning": "#faad14",
                  },
                }
              }
            >
              TalkVeganToMe
            </Text>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#E8E9EB",
            "height": 1,
          }
        }
      />
      <View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={null}
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#3371AA",
                "borderColor": "#bcbbc1",
                "flexDirection": "row",
                "padding": 14,
              }
            }
          >
            <View>
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "transparent",
                    "justifyContent": "center",
                  }
                }
                theme={
                  Object {
                    "colors": Object {
                      "disabled": "hsl(208, 8%, 90%)",
                      "divider": "#bcbbc1",
                      "error": "#ff190c",
                      "grey0": "#393e42",
                      "grey1": "#43484d",
                      "grey2": "#5e6977",
                      "grey3": "#86939e",
                      "grey4": "#bdc6cf",
                      "grey5": "#e1e8ee",
                      "greyOutline": "#bbb",
                      "platform": Object {
                        "android": Object {
                          "error": "#f44336",
                          "primary": "#2196f3",
                          "secondary": "#9C27B0",
                          "success": "#4caf50",
                          "warning": "#ffeb3b",
                        },
                        "ios": Object {
                          "error": "#ff3b30",
                          "primary": "#007aff",
                          "secondary": "#5856d6",
                          "success": "#4cd964",
                          "warning": "#ffcc00",
                        },
                      },
                      "primary": "#2089dc",
                      "searchBg": "#303337",
                      "secondary": "#8F0CE8",
                      "success": "#52c41a",
                      "warning": "#faad14",
                    },
                  }
                }
                underlayColor="white"
              >
                <Text
                  allowFontScaling={false}
                  style={
                    Array [
                      Object {
                        "color": null,
                        "fontSize": 24,
                      },
                      Object {
                        "backgroundColor": "transparent",
                        "color": "#E8E9EB",
                        "fontSize": 20,
                      },
                      Object {
                        "fontFamily": "Material Icons",
                        "fontStyle": "normal",
                        "fontWeight": "normal",
                      },
                      Object {},
                    ]
                  }
                  testID="iconIcon"
                >
                  Óóè
                </Text>
              </View>
            </View>
            <View
              width={16}
            />
            <View
              style={
                Object {
                  "flex": 1,
                  "justifyContent": "center",
                }
              }
            >
              <Text
                style={
                  Object {
                    "backgroundColor": "transparent",
                    "color": "#E8E9EB",
                    "fontSize": 20,
                  }
                }
                testID="listItemTitle"
                theme={
                  Object {
                    "colors": Object {
                      "disabled": "hsl(208, 8%, 90%)",
                      "divider": "#bcbbc1",
                      "error": "#ff190c",
                      "grey0": "#393e42",
                      "grey1": "#43484d",
                      "grey2": "#5e6977",
                      "grey3": "#86939e",
                      "grey4": "#bdc6cf",
                      "grey5": "#e1e8ee",
                      "greyOutline": "#bbb",
                      "platform": Object {
                        "android": Object {
                          "error": "#f44336",
                          "primary": "#2196f3",
                          "secondary": "#9C27B0",
                          "success": "#4caf50",
                          "warning": "#ffeb3b",
                        },
                        "ios": Object {
                          "error": "#ff3b30",
                          "primary": "#007aff",
                          "secondary": "#5856d6",
                          "success": "#4cd964",
                          "warning": "#ffcc00",
                        },
                      },
                      "primary": "#2089dc",
                      "searchBg": "#303337",
                      "secondary": "#8F0CE8",
                      "success": "#52c41a",
                      "warning": "#faad14",
                    },
                  }
                }
              >
                Test Parent
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "backgroundColor": "#E8E9EB",
              "height": 1,
            }
          }
        />
        <View
          animationType="slide"
          style={
            Object {
              "display": "none",
            }
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={null}
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#E8E9EB",
                  "borderBottomWidth": 0.5,
                  "borderColor": "#bcbbc1",
                  "flexDirection": "row",
                  "padding": 14,
                }
              }
            >
              <View
                style={
                  Object {
                    "flex": 1,
                    "justifyContent": "center",
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "backgroundColor": "transparent",
                      "fontSize": 17,
                    }
                  }
                  testID="listItemTitle"
                  theme={
                    Object {
                      "colors": Object {
                        "disabled": "hsl(208, 8%, 90%)",
                        "divider": "#bcbbc1",
                        "error": "#ff190c",
                        "grey0": "#393e42",
                        "grey1": "#43484d",
                        "grey2": "#5e6977",
                        "grey3": "#86939e",
                        "grey4": "#bdc6cf",
                        "grey5": "#e1e8ee",
                        "greyOutline": "#bbb",
                        "platform": Object {
                          "android": Object {
                            "error": "#f44336",
                            "primary": "#2196f3",
                            "secondary": "#9C27B0",
                            "success": "#4caf50",
                            "warning": "#ffeb3b",
                          },
                          "ios": Object {
                            "error": "#ff3b30",
                            "primary": "#007aff",
                            "secondary": "#5856d6",
                            "success": "#4cd964",
                            "warning": "#ffcc00",
                          },
                        },
                        "primary": "#2089dc",
                        "searchBg": "#303337",
                        "secondary": "#8F0CE8",
                        "success": "#52c41a",
                        "warning": "#faad14",
                      },
                    }
                  }
                >
                  Test Page
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</RCTSafeAreaView>
`;
